alf/environments/alf_gym3_wrapper.py

```diff
@@ -0,0 +1,415 @@
+        the convention.
+
+        Otherwise the reset will be ignored.
+
+        """
+        if self._support_force_reset:
+            self._gym3_env.act(np.array([-1] * self.batch_size))
+            self._done = [False] * self.batch_size
+        else:
+            logging.warning('reset() ignored by AlfGym3Wrapper')
+
+        reward, observation, _ = self._gym3_env.observe()
+
+        time_step = self._create_time_step(
+            reward,
+            observation,
+            step_type=[ds.StepType.FIRST] * self.batch_size,
+            # Faking actions
+            action=nest.map_structure(
+                lambda spec: spec.numpy_zeros(outer_dims=(self.batch_size, )),
+                self.action_spec()))
+
+        return time_step
+
+    # Implement abstract _step
+    def _step(self, action) -> ds.TimeStep:
+        """Implement ``_step()`` for compatibility
+
+        A very important note here is that special treatment is done at the end
+        of each episode. Unlike Gym environments, Gym3 environments do not
+        return ``done=True`` immediately. Instead, ``first=True`` is returned
+        upon the ``observe()`` of the NEXT ``act()``, which is actually the
+        FIRST FRAME of a new episode in Gym3 sense.
+
+        With this wrapper we CHANGED that definition, so that the frame with
+        ``first=True`` actually becomes the LAST FRAME of the previous episode.
+        Because the observation of the new episode can be DRAMATICALLY DIFFERENT
+        from the actual last frame of the previous episode, the previous
+        observation is returned in this case.
+
+        Therefore to summarize, effectively we will
+
+        1. Repeat the end-of-episode observation twice for each episode.
+
+        2. Throw away the first frame of each episode, and use the second frame
+           as if it is the first frame.
+
+        Gym3's official Gym wrapper DID THE SAME.
+
+        """
+        _, prev_observation, _ = self._gym3_env.observe()
+
+        np_action = nest.map_structure(lambda x: x.cpu().numpy(), action)
+        self._gym3_env.act(np_action)
+
+        reward, observation, first = self._gym3_env.observe()
+
+        # Override the obervation with the previous observation if that
+        # particular environment has ``first=True``.
+        for i in range(self.batch_size):
+            if first[i]:
+                observation[i] = prev_observation[i]
+
+        # This does the trick of repeating end-of-episode frames and throwing
+        # away first-of-episode frames.
+        step_type = [
+            ds.StepType.FIRST if d else
+            (ds.StepType.LAST if f else ds.StepType.MID)
+            for d, f in zip(self._done, first)
+        ]
+
+        time_step = self._create_time_step(
+            reward=reward,
+            observation=observation,
+            step_type=step_type,
+            action=action)
+
+        self._done = first
```

Comment:
Maybe self._done is not a good name anymore. Can be called `self._prev_first`.

Reply:
It was actually called `_prev_first` before renaming to `_done`. In our interpretation, the frame whose `first` is `true` is actually "the last frame of the previous episode", which means `done`. But I am open to the suggestion.

Reply:
Renamed back to `_prev_first`.